File: BI-TJV/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/table/TableActivity.java
package cz.wrzecond.restsys.table;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

import cz.wrzecond.restsys.R;
import cz.wrzecond.restsys.task.HttpMethod;
import cz.wrzecond.restsys.task.HttpRequest;
import cz.wrzecond.restsys.Activity;
import cz.wrzecond.restsys.Adapter;
import cz.wrzecond.restsys.CreateListener;
import cz.wrzecond.restsys.food.FoodActivity;
import cz.wrzecond.restsys.order.OrderFoodActivity;
import wrzecond.table.TjvTableReadDTO;

public class TableActivity extends Activity<TjvTableReadDTO> {

    private boolean loaded = false;

    @Override
    protected void onCreate (@Nullable Bundle si) {
        super.onCreate(si);
        super.overridePendingTransition(0, 0);

        loadNavigation(R.id.navigation_tables, R.layout.create_table);
        loaded = true;

        FloatingActionButton leftFab = findViewById(R.id.left_fab);
        leftFab.setImageResource(R.drawable.ic_inside_table);
        leftFab.setVisibility(View.VISIBLE);
        leftFab.setOnClickListener( v -> {
            Intent intent = new Intent(this, FoodActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
            intent.putExtra(FoodActivity.EXTRA_SELECTION_TABLE, -2);
            startActivityForResult(intent, FoodActivity.SELECTION_RESULT_CODE);
        } );

        FloatingActionButton middleFab = findViewById(service.isAdmin() ? R.id.middle_fab : R.id.fab);
        middleFab.setImageResource(R.drawable.ic_outside_table);
        middleFab.setVisibility(View.VISIBLE);
        middleFab.setOnClickListener( v -> {
            Intent intent = new Intent(this, FoodActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
            intent.putExtra(FoodActivity.EXTRA_SELECTION_TABLE, -3);
            startActivityForResult(intent, FoodActivity.SELECTION_RESULT_CODE);
        } );
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (loaded)
            loadList();
    }

    @Override
    protected String getEndpoint () {
        return "/tables/";
    }

    @Override
    protected Type getDtoListType() {
        return new TypeToken<List<TjvTableReadDTO>>(){}.getType();
    }

    @Override
    protected Adapter<TjvTableReadDTO> createAdapter (List<TjvTableReadDTO> dtos) {
        return new TableAdapter(this, dtos);
    }

    @Override
    protected Adapter<TjvTableReadDTO> createAdapter (@StringRes int errorResId) {
        return new TableAdapter(this, errorResId);
    }

    @Override
    protected RecyclerView.LayoutManager getLayoutManager (boolean error) {
        return error ? super.getLayoutManager(true) : new GridLayoutManager(this, 4);
    }

    @Override
    public CreateListener<TjvTableReadDTO> createListener (AlertDialog dialog, TjvTableReadDTO dto) {
        dialog.setTitle(dto != null ? getString(R.string.table_edit, dto.getId()) : getString(R.string.table_new));
        return new TableCreateListener(this, dialog, dto);
    }

    @Override
    protected void onActivityResult (int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // Create new order
        if (requestCode == FoodActivity.SELECTION_RESULT_CODE && resultCode == RESULT_OK && data != null) {
            ProgressBar progressBar = findViewById(R.id.list_loading);
            int tableId = data.getIntExtra(FoodActivity.EXTRA_SELECTION_TABLE, -1);
            progressBar.setVisibility(View.VISIBLE);

            String url = ( tableId >= 0 ? "/order-food/at/" + tableId : (
                tableId == -2 ? "/order-food/type/INSIDE" : "/order-food/type/OUTSIDE"
            ) );

            // Send request
            HttpRequest request = new HttpRequest(url, HttpMethod.POST);
            request.setBody(data.getStringExtra(FoodActivity.EXTRA_SELECTION));
            service.scheduleTask( response -> {
                // Failure
                if ( !response.getStatus().isSuccess() ) {
                    progressBar.setVisibility(View.GONE);
                    Toast.makeText(this, getString(R.string.table_order_create_failure,
                            response.getStatus().name()), Toast.LENGTH_SHORT).show();
                    return;
                }

                if (tableId < 0) {
                    Toast.makeText(this, R.string.table_order_create_success, Toast.LENGTH_SHORT).show();
                    loadList();
                    return;
                }

                // Launch order view
                Intent intent = new Intent(this, OrderFoodActivity.class);
                intent.putExtra(OrderFoodActivity.EXTRA_TABLE_ID, tableId);
                intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
                startActivity(intent);
            }, request );
        }
    }

}

File: BI-KOT/semestral/rest-client/app/src/main/java/cz/wrzecond/restsys/table/TableActivity.kt
package cz.wrzecond.restsys.table

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.gson.reflect.TypeToken
import cz.wrzecond.restsys.Activity
import cz.wrzecond.restsys.R
import cz.wrzecond.restsys.food.FoodActivity
import cz.wrzecond.restsys.order.OrderFoodActivity
import cz.wrzecond.restsys.task.HttpMethod
import cz.wrzecond.restsys.task.HttpRequest
import wrzecond.TjvTableReadDTO
import java.lang.reflect.Type

class TableActivity : Activity<TjvTableReadDTO>() {

    // Properties
    private var loaded = false

    override fun onCreate (si: Bundle?) {
        super.onCreate(si)
        overridePendingTransition(0, 0)

        // Init navigation
        loadNavigation(R.id.navigation_tables, R.layout.create_table)
        loaded = true

        // Order on all tables inside
        val leftFab = findViewById<FloatingActionButton>(R.id.left_fab)
        leftFab.setImageResource(R.drawable.ic_inside_table)
        leftFab.visibility = View.VISIBLE
        leftFab.setOnClickListener {
            val intent = Intent(this, FoodActivity::class.java)
                .addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
                .putExtra(FoodActivity.EXTRA_SELECTION_TABLE, TABLE_ID_SPECIAL_ALL_INSIDE)
            startActivityForResult(intent, FoodActivity.SELECTION_RESULT_CODE)
        }

        // Order on all tables outside
        val middleFab = findViewById<FloatingActionButton>(R.id.middle_fab)
        middleFab.setImageResource(R.drawable.ic_outside_table)
        middleFab.visibility = View.VISIBLE
        middleFab.setOnClickListener {
            val intent = Intent(this, FoodActivity::class.java)
                    .addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
                    .putExtra(FoodActivity.EXTRA_SELECTION_TABLE, TABLE_ID_SPECIAL_ALL_OUTSIDE)
            startActivityForResult(intent, FoodActivity.SELECTION_RESULT_CODE)
        }
    }

    override fun onResume() {
        super.onResume()
        if (loaded)
            loadList()
    }

    override val endpoint : String
        get() = "/tables/"

    override val dtoListType: Type
        get() = object : TypeToken<List<TjvTableReadDTO>>(){}.type

    override fun createAdapter (dtos: List<TjvTableReadDTO>)
        = TableAdapter (this, dtos = dtos)

    override fun createAdapter (errorResId: Int)
        = TableAdapter (this, errorResId = errorResId)

    override fun getLayoutManager (error: Boolean) : RecyclerView.LayoutManager
        = if (error) super.getLayoutManager(error) else GridLayoutManager(this, 4)

    override fun createListener (dialog: AlertDialog, dto: TjvTableReadDTO?)
        = TableCreateListener (this, dialog, dto)
            .apply {
                dialog.setTitle(dto?.let { getString(R.string.table_edit, it.id) } ?:
                getString(R.string.table_new))
            }

    override fun onActivityResult (requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        // Not interested
        if (requestCode != FoodActivity.SELECTION_RESULT_CODE || resultCode != RESULT_OK || data == null)
            return

        // Show progress
        val progressBar = findViewById<ProgressBar>(R.id.list_loading)
        progressBar.visibility = View.VISIBLE

        // Init data
        val tableId = data.getIntExtra(FoodActivity.EXTRA_SELECTION_TABLE, -1)
        val url = when (tableId) {
            TABLE_ID_SPECIAL_ALL_INSIDE  -> "/order-food/type/INSIDE"
            TABLE_ID_SPECIAL_ALL_OUTSIDE -> "/order-food/type/OUTSIDE"
            else -> "/order-food/at/$tableId"
        }

        // Schedule request
        val request = HttpRequest(url, HttpMethod.POST, body = data.getStringExtra(FoodActivity.EXTRA_SELECTION))
        service.scheduleTask(request) { response ->
            // Creation failure
            if (!response.status.isSuccess) {
                progressBar.visibility = View.GONE
                Toast.makeText(this, getString(R.string.table_order_create_failure,
                        response.status.name), Toast.LENGTH_SHORT).show()
                return@scheduleTask
            }

            // Creation (at specific table type) success
            if (tableId == TABLE_ID_SPECIAL_ALL_INSIDE || tableId == TABLE_ID_SPECIAL_ALL_OUTSIDE) {
                Toast.makeText(this, R.string.table_order_create_success, Toast.LENGTH_SHORT).show()
                loadList()
                return@scheduleTask
            }

            // Creation (at specific table) success = show order view
            val intent = Intent(this, OrderFoodActivity::class.java)
                .addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
                .putExtra(OrderFoodActivity.EXTRA_TABLE_ID, tableId)
            startActivity(intent)
        }
    }

    companion object {
        private const val TABLE_ID_SPECIAL_ALL_INSIDE  = -2
        private const val TABLE_ID_SPECIAL_ALL_OUTSIDE = -3
    }

}